### DISCUSSION TITLE
Deep Merge two instances

### DISCUSSION THREAD

**Discussion by luwol03** [pydantic/pydantic#4352] [Question] [answered]
Hello,

how can I merge two pydantic instances but keep the objects. Lets say:
```py
class User(BaseModel):
    name: Optional[str]
    role: Optional[str]
    friends: Optional[List['User']] = []


blacky = User(name="blacky")
peter = User(role="AB", friends=[blacky])

bob = User(name="bob")
shawn = User(name="shawn", friends=[bob])

print("Before", type(peter.friends[0]), peter.friends) # Before <class '__main__.User'> [User(name='blacky', role=None, friends=[])]
peter = peter.copy(update=shawn.dict(exclude_unset=True), deep=True)
print("AFTER", type(peter.friends[0]), peter.friends) # AFTER <class 'dict'> [{'name': 'bob'}]
# Expected: <class '__main__.User'> [User(name='blacky', role=None, friends=[]), User(name='bob', role=None, friends=[])]
``` 
This method is indeed merging `peter` and `shawn` but its destroying their reference, as they are first converted to a dict with `.dict(exclude_unset=True)`. And at the end the friend is of type `<class 'dict'>` instead of `<class '__main__.User'>`.

So my question is, how can I deep merge two instances without a reserialization and also keep the references.

luwol03

**Comment by samuelcolvin** [comment]
We don't have a method for this, you'll have to write your own logic to
achieve this.

**Reply by luwol03** [reply]
I'm new to pydantic, can you please give me some hints how I can archive that?

**Reply by samuelcolvin** [reply]
There's nothing specific about pydantic. It's just getting and setting attributes on classes, same as vanilla python.

I'm on my phone so can't provide code to help, sorry.

**Reply by luwol03** [reply]
Thank you for your quick replays, awesome.

There is no rush, I would really appreciate it, if you could provide a code example how such a merge method could look like.

**Edit:** I'm searching for a general approach by looping over all available fields and their values. But I have no idea how to do that.

**Reply by luwol03** [reply, marked as answer]
I found it, `__fields__` was the missing part. What are your thoughts about that?

```py
def merge_objects(a, b, clone=False):
    if clone:
        a = a.copy()

    for k in b.__fields__.keys():
        a_attr = getattr(a, k)
        b_attr = getattr(b, k)

        if b_attr is None:
            continue

        if a_attr is None:
            setattr(a, k, b_attr)
            continue

        if isinstance(a_attr, list) and isinstance(b_attr, list):
            for i in b_attr:
                if i not in a_attr:
                    a_attr.append(i)

        else:
            setattr(a, k, b_attr)

    return a
```
